cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,327 @@
+
+        Args:
+            circuit: Input circuit to the new transformer stage.
+            transformer_name: Name of the new transformer stage.
+        """
+        if self._stack:
+            self._logs[self._stack[-1]].nested_loggers += (self._curr_id,)
+        self._logs.append(LoggerNode(self._curr_id, transformer_name, circuit, circuit))
+        self._stack.append(self._curr_id)
+        self._curr_id += 1
+
+    def log(self, *args: str, level: LogLevel = LogLevel.INFO) -> None:
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+
+        Raises:
+            ValueError: If there's no active transformer on the stack.
+        """
+        if len(self._stack) == 0:
+            raise ValueError('No active transformer found.')
+        self._logs[self._stack[-1]].logs += ((level, args),)
+
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+
+        Raises:
+            ValueError: If `transformer_name` is different from currently active transformer name.
+        """
+        tid = self._stack.pop()
+        if self._logs[tid].transformer_name != transformer_name:
+            raise ValueError(
+                f"Expected `register_final` call for currently active transformer "
+                f"{self._logs[tid].transformer_name}."
+            )
+        self._logs[tid].final_circuit = circuit
+
+    def show(self, level: LogLevel = LogLevel.INFO) -> None:
+        """Show the stored logs >= level in the desired format."""
+
+        def print_log(log: LoggerNode, pad=''):
+            print(pad, f"Transformer-{1+log.transformer_id}: {log.transformer_name}", sep='')
+            print(pad, "Initial Circuit:", sep='')
+            print(textwrap.indent(str(log.initial_circuit), pad), "\n", sep='')
+            for log_level, log_text in log.logs:
+                if log_level.value >= level.value:
+                    print(pad, log_level, *log_text)
+            print("\n", pad, "Final Circuit:", sep='')
+            print(textwrap.indent(str(log.final_circuit), pad))
+            print("----------------------------------------")
+
+        done = [0] * self._curr_id
+
+        def dfs(log: LoggerNode, pad=''):
+            print_log(log, pad)
+            done[log.transformer_id] = True
+            for child_id in log.nested_loggers:
+                dfs(self._logs[child_id], pad + ' ' * 4)
+
+        for i in range(self._curr_id):
+            if not done[i]:
+                dfs(self._logs[i])
```

Comment:
Can we swap this over to the iterative version ? I personally get pretty worried seeing recursion in library code unless it's absolutely needed.